File: src_server/webstuff/routechgs.py
import functools
import itertools
import re
from datetime import datetime
import cherrypy
import sys

sys.path.append('../') # i hate python so much

from load_service_alerts import parse_unixtime_into_jerusalem_tz, JERUSALEM_TZ, USE_CASE
from junkyard import remove_all_occurrences_from_list

def find_representative_date_for_route_changes_in_alert(alert):
    _active_periods_parsed = map(
        lambda period: list(map(parse_unixtime_into_jerusalem_tz, period)),
        alert["active_periods"]["raw"]
    )

    TODAY_IN_JERUS = JERUSALEM_TZ \
        .fromutc(datetime.utcnow())\
        .replace(
            hour=0,
            minute=0,
            second=0,
            microsecond=0
        )
    
    representative_date = None

    if alert["is_expired"]:
        for start, end in _active_periods_parsed:
            if end is None:
                representative_date = TODAY_IN_JERUS
                break
            
            if representative_date is None or end > representative_date:
                representative_date = end
    elif alert["is_deleted"]:
        representative_date = alert["last_end_time"].replace(hour=0, minute=0, second=0, microsecond=0)
    #   TODO representative_date = deleted_date
    else:
        for start, end in _active_periods_parsed:
            hasStart = start is not None
            hasEnd   = end is not None

            if not hasEnd and not hasStart:
                # unbounded period - use today
                representative_date = TODAY_IN_JERUS
                break
            
            if hasEnd and end <= TODAY_IN_JERUS:
                # period ended already, skip it
                continue
            
            if not hasStart or start <= TODAY_IN_JERUS:
                # period is active now!
                representative_date = TODAY_IN_JERUS
                break

            # period is in the future

            if representative_date is None or start < representative_date:
                representative_date = start
    
    return representative_date # or TODAY_IN_JERUS?????





# am i gonna understand this regex in a week? let's find out!
ROUTE_DESC_DIR_ALT_PATTERN = re.compile(r'^[^-]+-([^-]+)-([^-]+)$')


def label_line_changes_headsigns_for_direction_and_alternative(line_changes):
    """for chg in line_changes: adds (direction 1, alternative 2) labels to duplicate headsigns,
    and deletes route_desc from the dictionary"""

    dict_headsign_to_dir_alt_pairs = {}

    for chg in line_changes:
        headsign = chg["to_text"]
        route_desc = chg["route_desc"]
        del chg["route_desc"]

        if headsign not in dict_headsign_to_dir_alt_pairs:
            dict_headsign_to_dir_alt_pairs[headsign] = []
        
        dir_alt_pair = ROUTE_DESC_DIR_ALT_PATTERN.findall(route_desc)[0]
        dict_headsign_to_dir_alt_pairs[headsign].append(dir_alt_pair)
        chg["dir_alt_pair"] = dir_alt_pair
    
    ordered_dir_alt_namepairs = label_headsigns_for_direction_and_alternative(
        dict_headsign_to_dir_alt_pairs,
        dict_headsign_to_dir_alt_pairs,
        map(
            lambda x: [x["to_text"], x["dir_alt_pair"]],
            line_changes
        )
    )

    for chg, namepair in zip(line_changes, ordered_dir_alt_namepairs):
        del chg["dir_alt_pair"]
        chg["dir_name"] = namepair[0]
        chg["alt_name"] = namepair[1]

def label_headsigns_for_direction_and_alternative(
    alt_dict_headsign_to_dir_alt_pairs,
    dir_dict_headsign_to_dir_alt_pairs,
    headsigns_with_diralts,
    label_dirs_per_alt=False
):

    def per_headsign(h):
        headsign, dir_alt_pair = h
        alt_dups = alt_dict_headsign_to_dir_alt_pairs[headsign]
        dir_dups = dir_dict_headsign_to_dir_alt_pairs[headsign]

        dir_name = None
        alt_name = None

        if len(alt_dups) == 1 and len(dir_dups) == 1:
            # no duplicates for this headsign! yay upside down smiley
            return [dir_name, alt_name]
        
        dir_id, alt_id = dir_alt_pair

        if any(map(lambda x: x[0] != dir_id and (not label_dirs_per_alt or x[1] == alt_id), dir_dups)):
            # if there's any dups with a different direction id

            # in some distant nebulous future, i could try giving actual names
            # to the directions and alternatives; but not today, not quite yet
            # bukra fil mishmish
            dir_name = str(
                # possibly the slowest most inefficient way to do this but as
                # stated earlier, yours truly is truly bad at algo
                sorted(set([d for d, a in dir_dups if (not label_dirs_per_alt or a == alt_id)])).index(dir_id) + 1
            )
        
        if alt_id not in ['#', '0'] and any(map(lambda x: x[1] != alt_id, alt_dups)):
            # if there's any dups with a different alternative id
            # (and also this isn't the main alternative)

            # remember what i said about actual names for directions?
            # well, same for the alternatives
            # i mean, like, how the heck does one even approach this problem???
            # we're given basically zero computer readable information that
            # summarizes the differences between directions/alternatives!
            # i guess if someone was eager enough, they COULD go through the
            # stop_seq of a representative trip, but then, how do you turn that
            # into not only user-readable info, but user-useful info, that isn't
            # too long to fit in the ui?????? can't just dump stop names/city names,
            # and detecting street names would be absolutely hellish
            
            # so uhm,, yeah
            # numbers it is for now

            # again possibly the slowest most inefficient blah blah blah
            # note: if a not in ['#', '0'] cause we don't care about the main alternative here
            # i want to display: Towards A, Towards A (Alt 1), Towards A (Alt 2)
            # and not quite:     Towards A, Towards A (Alt 2), Towards A (Alt 3)
            alternatives = sorted(set([a for d, a in alt_dups if a not in ['#', '0']]))

            if len(alternatives) == 1:
                # but also we want to display: Towards A, Towards A (Alt)
                # and not qutie:               Towards A, Towards A (Alt 1)
                # because "1" doesn't make sense when there's just the one
                alt_name = "#"
            else:
                alt_name = str(alternatives.index(alt_id) + 1)
        
        return (dir_name, alt_name)
    
    return map(per_headsign, headsigns_with_diralts)

def bounding_box_for_stops(stop_ids, stops_for_map):
    """get bounding box of affected stops, for setting the map's bounding box"""

    map_bounding_box = {
        "min_lon": None,
        "min_lat": None,
        "max_lon": None,
        "max_lat": None
    }

    for stop_id in stop_ids:
        try:
            stop = stops_for_map[stop_id]
        except KeyError:
            continue

        lon = stop["stop_lon"]
        lat = stop["stop_lat"]

        if map_bounding_box["min_lon"] is None or map_bounding_box["min_lon"] > lon:
            map_bounding_box["min_lon"] = lon
        if map_bounding_box["min_lat"] is None or map_bounding_box["min_lat"] > lat:
            map_bounding_box["min_lat"] = lat
        if map_bounding_box["max_lon"] is None or map_bounding_box["max_lon"] < lon:
            map_bounding_box["max_lon"] = lon
        if map_bounding_box["max_lat"] is None or map_bounding_box["max_lat"] < lat:
            map_bounding_box["max_lat"] = lat
    
    return map_bounding_box

def compute_stop_ids_incl_adj_for_single_route_change(alert_minimal, orig_stop_seq):

    stop_ids = set()

    first = True
    prev_stop_id    = None
    prev_is_deleted = None
    for stop_id in orig_stop_seq:
        is_deleted = stop_id in alert_minimal["deleted_stop_ids"]

        if is_deleted:
            # deleted stops need to be in the bounding box
            stop_ids.add(stop_id)

            if not first and not prev_is_deleted:
                # and also non-deleted stops that come right before a deleted stop
                stop_ids.add(prev_stop_id)
        elif not first and prev_is_deleted:
            # and also also non-deleted stops that come right after a new stop
            stop_ids.add(stop_id)
        
        first = False
        prev_stop_id = stop_id
        prev_is_deleted = is_deleted

    first = True
    prev_stop_id = None
    prev_is_new  = None
    for stop_id, is_new in alert_minimal["updated_stop_sequence"]:
        if is_new:
            # new stops need to be in the bounding box
            stop_ids.add(stop_id)
        
            if not first and not prev_is_new:
                # and also non-new stops that come right before a new stop
                stop_ids.add(prev_stop_id)
        elif not first and prev_is_new:
            # and also also non-new stops that come right after a new stop
            stop_ids.add(stop_id)
    
        first = False
        prev_stop_id = stop_id
        prev_is_new  = is_new
    
    if not len(stop_ids):
        # sometimes the route changes don't actually have any changes????
        # i have no idea what the fuck they're doing over there in the wherever
        # they're doing whatever ther fuck it is they're doing
        return set(map(lambda x: x[0], alert_minimal["updated_stop_sequence"]))
    else:
        return stop_ids

def list_of_alerts_to_active_period_intersections_and_bitmasks(alerts_minimal):
    all_active_period_boundaries = []

    for idx, alert_minimal in enumerate(alerts_minimal):
        for start, end in alert_minimal["active_periods"]["raw"]:
            all_active_period_boundaries.append((start or 0, idx, False))
            all_active_period_boundaries.append((end or 7258118400, idx, True)) # alerts with no specified end time get the timestamp for 2200-01-01 00:00 UTC
    
    all_active_period_boundaries.sort()

    all_periods = []
    current_period = None

    while len(all_active_period_boundaries):
        timestamp, idx, is_end = all_active_period_boundaries.pop(0)

        if not current_period:
            # this is the first period boundary we're encountering

            if is_end:
                # i don't know what kind of terrible data would get us HERE
                # but if it does, ignore it and hope for the best lol!
                continue
            
            current_period = {"start": timestamp, "end": None, "bitmask": 0}
            all_periods.append(current_period)
        elif current_period["start"] != timestamp:
            current_period["end"] = timestamp
            current_period = {"start": timestamp, "end": None, "bitmask": current_period["bitmask"]}
            all_periods.append(current_period)
        
        idx_bitmask = 1 << idx
        if not is_end:
            # this alert just started, add it to the bitmask
            current_period["bitmask"] |= idx_bitmask
        elif current_period["bitmask"] & idx_bitmask:
            # this alert just ended, and it actually was in the bitmask
            # (i could theoretically do &= ~idx_bitmask, but python's bitwise
            # not feels unpredictable, so for my own sake, i added the extra
            # check and did a xor instead)
            current_period["bitmask"] ^= idx_bitmask
    
    return all_periods



def does_alert_have_route_changes(alert):
    return alert["use_case"] in [
        USE_CASE.STOPS_CANCELLED.value,
        USE_CASE.ROUTE_CHANGES_FLEX.value,
        USE_CASE.ROUTE_CHANGES_SIMPLE.value
    ]


def apply_alert_to_route(
    alert,
    route_id,
    gtfsdbapi,
    representative_date = None,
    representative_trip_id = None,
    raw_stop_seq = None,
    updated_stop_seq = None,
    deleted_stop_ids = None,
    mut_all_stop_ids_set = None
):
    # TODO: it looks like i never took care of the REGION use case lmaooooooo
    #       i'd feel much more comfortable implementing it if they,,, uh,,,,,,,,,,,, ever used it :|
    #       but sure; i can try to do it al iver just in case
    if not does_alert_have_route_changes(alert):
        return None

    # if needed, get a representative date
    if not representative_date and not representative_trip_id and not updated_stop_seq:
        representative_date = find_representative_date_for_route_changes_in_alert(alert)
    
    # if needed, get a representative trip
    if not representative_trip_id and not updated_stop_seq:
        representative_trip_id = gtfsdbapi.get_representative_trip_id(route_id, representative_date)
    
    # and if needed, get that trip's stop sequence
    if not updated_stop_seq:
        if not raw_stop_seq:
            raw_stop_seq = gtfsdbapi.get_stop_seq(representative_trip_id)
        
        if mut_all_stop_ids_set:
            mut_all_stop_ids_set.update(raw_stop_seq)
        
        updated_stop_seq = [
            (stop_id, False) # (stop_id, is_added)
            for stop_id in raw_stop_seq
        ]
        deleted_stop_ids = []
    
    # and actually do the magic of computing the new stop sequence
    if alert["use_case"] == USE_CASE.STOPS_CANCELLED.value:
        # special case :|
        for removed_stop_id in alert["removed_stop_ids"]:
            times_removed = remove_all_occurrences_from_list(
                # TODO: support removing a stop thats been inserted by an alert
                updated_stop_seq,
                (removed_stop_id, False)
            )

            if times_removed > 0 or len(alert["relevant_route_ids"]) == 1:
                deleted_stop_ids.append(removed_stop_id)
    else:
        changes_for_route = alert["schedule_changes"][route_id]

        for change in changes_for_route:
            if "removed_stop_id" in change:
                times_removed = remove_all_occurrences_from_list(
                    # TODO: support removing a stop thats been inserted by an alert
                    updated_stop_seq,
                    (change["removed_stop_id"], False)
                )

                if times_removed == 0:
                    cherrypy.log(f"tried removing stop that's not on a route; route_id={route_id}, {repr(change)}, alert_id={alert['id']}, trip_id={representative_trip_id})")
                
                if times_removed > 0 or len(alert["relevant_route_ids"]) == 1:
                    deleted_stop_ids.append(change["removed_stop_id"])
            elif "added_stop_id" in change:
                dest_idx = None
                for idx, t in enumerate(updated_stop_seq):
                    # can't JUST search for (relative_stop_id, False) because the
                    # relative_stop_id might have been added some previous iteration
                    if t[0] == change["relative_stop_id"]:
                        dest_idx = idx
                        break
                    # nice lil edge case the mot didn't think about:
                    # what if a trip stops somewhere twice, and we're told to add
                    # another stop before/after that one that appears twice?

                    # should i like check for that edge case? and put the stop.....
                    # uhm.... where.... the .... distance to the other stops?
                    # is shortest? idk; or i'll just bug, and blame the government
                    # because that's easier
                
                if dest_idx is None:
                    # didn't find the stop we're supposed to add relative to
                    cherrypy.log(
                        f"tried adding stop relative to stop not on route; route_id={route_id}, {repr(change)}, alert_id={alert['id']}, trip_id={representative_trip_id}"
                    )
                    continue

                if not change["is_before"]:
                    dest_idx += 1
                
                updated_stop_seq.insert(dest_idx, (change["added_stop_id"], True))
                cherrypy.log(f"added stop {change['added_stop_id']} to route {route_id} at index {dest_idx}")

                if mut_all_stop_ids_set:
                    mut_all_stop_ids_set.add(change["added_stop_id"])
    
    if mut_all_stop_ids_set:
        mut_all_stop_ids_set.update(deleted_stop_ids)

    # aaahahahahahahahha in like 2 weeks im gonna look at this return statemnet
    # and be so perplexed lmao sgonna be real funny and by funny i mean ill get
    # [REDACTED] again ahahahahah
    return {
        "alert": alert,
        "route_id": route_id,
        "gtfsdbapi": gtfsdbapi,
        "representative_date": representative_date,
        "representative_trip_id": representative_trip_id,
        "raw_stop_seq": raw_stop_seq,
        "updated_stop_seq": updated_stop_seq,
        "deleted_stop_ids": deleted_stop_ids,
        "mut_all_stop_ids_set": mut_all_stop_ids_set
    }

File: src_node/webstuff/routeChgs.ts
import { DateTime } from "luxon";
import { AlertUseCase, AlertWithRelatedInDb } from "../dbTypes.js";
import { AlertsDbApi } from "./alertsDbApi.js";
import { GtfsDbApi } from "./gtfsDbApi.js";
import { findRepresentativeDateForRouteChangesInAlert, getHeadsign } from "./alerts.js";
import winston from "winston";
import { MapBoundingBox, RouteChangeForApi, StopForMap } from "../apiTypes.js";
import { chainIterables, compareNple, inPlaceSortAndUnique, zip } from "../generalJunkyard.js";

export async function getRouteChanges(
    alertId: string,
    alertRaw: AlertWithRelatedInDb|null,
    alertsDbApi: AlertsDbApi,
    gtfsDbApi: GtfsDbApi
) {
    // 1. get alert and relevant routes, if no route changes return null

    alertRaw = alertRaw ?? await alertsDbApi.getSingleAlert(alertId);

    if (!alertRaw) {
        throw new Error(`Could not computer RouteChanges: alert with id ${alertId} not found`);
    }

    if (!doesAlertHaveRouteChanges(alertRaw)) {
        return null;
    }

    const changes_by_agency_and_line: Record<string, Record<string, RouteChangeForApi[]>> = {};

    const allStopIds = new Set([
        ...alertRaw.added_stop_ids,
        ...alertRaw.removed_stop_ids
    ]);

    const nearStopIds = new Set<string>([]);

    let representativeDate: DateTime|null = null;

    for (const route_id of alertRaw.relevant_route_ids) {
        // 2. actually apply the changes in the alert to this route
        const updates: ApplyAlertState|null = await applyAlertToRoute(
            gtfsDbApi,
            alertRaw,
            route_id,
            allStopIds,
            representativeDate
        );

        if (!updates) {
            continue; // this "shouldn't" happen anyway lol
        }

        const stopSeq = updates.updatedStopSeq;
        const representativeTripId = updates.representativeTripId ?? ""; // this "shouldn't" be null in this case lol
        const rawStopSeq = updates.rawStopSeq ?? []; // this "shouldn't" either haha
        const deletedStopIds = updates.deletedStopIds;
        representativeDate = updates.representativeDate;

        // 3. for the map bounding box, collect all stop_ids of stop that
        // are adjacent to added stops

        for (let i = 1; i < stopSeq.length; i++) {
            const [stopId, isAdded] = stopSeq[i] ?? ["", false];
            const [prevStopId, prevStopIsAdded] = stopSeq[i - 1] ?? ["", false];

            if (isAdded && !prevStopIsAdded) {
                nearStopIds.add(prevStopId);
            } else if (!isAdded && prevStopIsAdded) {
                nearStopIds.add(stopId);
            }
        }

        // 4. decide what to call each route???? as in from/to????? aaaaa israel
        //    (and get route_short_name and agency_id)

        const routeMetadata = (await gtfsDbApi.getRouteMetadata([route_id]))[route_id];

        if (!routeMetadata) {
            throw new Error("Couldn't find details for route with id " + route_id);
        }

        const to_text = await getHeadsign(representativeTripId, rawStopSeq, gtfsDbApi);

        // 5. get shape

        const shape = await gtfsDbApi.getShapePoints(representativeTripId)
            ?? await getFallbackShape(rawStopSeq, gtfsDbApi); // straight lines if we couldn't find a shape
        
        const routeChangeData = {
            ...routeMetadata,
            to_text,
            shape,
            deleted_stop_ids: [...deletedStopIds],
            updated_stop_sequence: stopSeq
        };

        // add to the result struct
        const agencyLines = changes_by_agency_and_line[routeChangeData.agency_id]
            ?? (changes_by_agency_and_line[routeChangeData.agency_id] = {});
        
        const lineChanges = agencyLines[routeChangeData.line_number]
            ?? (agencyLines[routeChangeData.line_number] = []);
        
        lineChanges.push(routeChangeData);

        winston.debug("done processing route_id " + route_id);
    }

    // --> bonus step cause i'm thorough, and motivated by hatred and spite:
    //     sort out any duplicate to_text
    for (const agencyLines of Object.values(changes_by_agency_and_line)) {
        for (const lineChanges of Object.values(agencyLines)) {
            labelLineChangesHeadsignsForDirectionAndAlternative(lineChanges);
        }
    }

    // 6. get all stops' metadata
    const stops_for_map = await gtfsDbApi.getStopsForMap([...allStopIds]);

    // 7. sort each line's changes by.... uhhhh..... good question le'ts
    //    decide on this issue randomly lmao
    for (const agencyLines of Object.values(changes_by_agency_and_line)) {
        for (const lineChanges of Object.values(agencyLines)) {
            lineChanges.sort(
                (a, b) => compareNple(
                    [a.to_text, a.dir_name ?? "", a.alt_name ?? ""],
                    [b.to_text, b.dir_name ?? "", b.alt_name ?? ""]
                )
                // other candidates:
                //  - always north->south/west->east before opposite?
                //  - always big place to small place before opposite?
                //  - by gtfs direction_id???
                //  - by mot route license id thing (route_desc)
                //  - random order for maximum fun! party horn emoji!
            );
        }
    }

    // 8. bounding box for the map widget
    const map_bounding_box = boundingBoxForStops(
        chainIterables(
            alertRaw.added_stop_ids,
            alertRaw.removed_stop_ids,
            nearStopIds
        ),
        stops_for_map
    );

    return {
        route_changes: changes_by_agency_and_line,
        stops_for_map,
        map_bounding_box
    };
}

function doesAlertHaveRouteChanges(alertRaw: AlertWithRelatedInDb) {
    return (
        alertRaw.use_case === AlertUseCase.StopsCancelled
        || alertRaw.use_case === AlertUseCase.RouteChangesFlex
        || alertRaw.use_case === AlertUseCase.RouteChangesSimple
        // TODO region use case???
    );
}

type ApplyAlertState = {
    representativeDate: DateTime|null,
    representativeTripId: string|null,
    rawStopSeq: string[]|null,
    updatedStopSeq: [string, boolean][],
    deletedStopIds: Set<string>
};

async function applyAlertToRoute(
    gtfsDbApi: GtfsDbApi,
    alertRaw: AlertWithRelatedInDb,
    route_id: string,
    mutAllStopIdsSet: Set<string>|null = null,
    representativeDate: DateTime|null = null,
    representativeTripId: string|null = null,
    rawStopSeq: string[]|null = null,
    updatedStopSeq: [string, boolean][]|null = null,
    deletedStopIds: Set<string>|null = null
): Promise<ApplyAlertState|null> {
    // TODO: it looks like i never took care of the REGION use case lmaoooooo
    //       i'd feel much more comfortable implementing it if they,, uh,,,,,, ever used it :|
    //       but sure; i can try to do it al iver just in case


    if (!doesAlertHaveRouteChanges(alertRaw)) {
        return null;
    }
    
    if (!updatedStopSeq) {
        // to create an updatedStopSeq, we first need to get a representative trip

        if (!representativeTripId) {
            // to get a representative trip, we first need to get a representative date

            if (!representativeDate) {
                representativeDate = findRepresentativeDateForRouteChangesInAlert(alertRaw);
            }

            representativeTripId = await gtfsDbApi.getRepresentativeTripId(route_id, representativeDate);
        }

        if (!representativeTripId) {
            throw new Error(`Could not find a representative trip id for route ${route_id} on date ${representativeDate?.toFormat('yyyy-MM-dd')}`);
        }

        if (!rawStopSeq) {
            rawStopSeq = await gtfsDbApi.getStopSeq(representativeTripId);
        }

        if (mutAllStopIdsSet) {
            rawStopSeq.forEach(s => mutAllStopIdsSet.add(s));
        }

        updatedStopSeq = rawStopSeq.map(
            stop_id => [stop_id, false] // [stop_id, is_added]
        );
        deletedStopIds = new Set();
    }

    if (!deletedStopIds) {
        deletedStopIds = new Set();
    }

    // and actually do the magic of computing the new stop sequence
    if (alertRaw.use_case === AlertUseCase.StopsCancelled) {
        // special case :|

        for (const removedStopId of alertRaw.removed_stop_ids) {
            const didRemove = removeStopFromUpdatedStopSeq(updatedStopSeq, removedStopId);

            if (didRemove || alertRaw.relevant_route_ids.length === 1) {
                deletedStopIds.add(removedStopId);
            }
        }
    } else if (alertRaw.use_case === AlertUseCase.RouteChangesFlex || alertRaw.use_case === AlertUseCase.RouteChangesSimple) {
        const changesForRoute = alertRaw.schedule_changes?.[route_id];
        if (!changesForRoute) return null;

        for (const change of changesForRoute) {
            if ("removed_stop_id" in change) {
                const didRemove = removeStopFromUpdatedStopSeq(updatedStopSeq, change.removed_stop_id);

                // the old python server used to warn about when a stop couldn't
                // actually be found in the updatedStopSeq, but uh,,, that happens
                // so often that it's just.... useless to even care about it *facepalm*

                if (didRemove || alertRaw.relevant_route_ids.length === 1) {
                    deletedStopIds.add(change.removed_stop_id);
                }
            } else {
                let destIdx = updatedStopSeq.findIndex(([s, _]) => s === change.relative_stop_id);

                // nice lil edge case the mot didn't think about:
                // what if a trip stops somewhere twice, and we're told to add
                // another stop before/after that one that appears twice?

                // should i like check for that edge case? and put the stop.....
                // uhm.... where.... the .... distance to the other stops?
                // is shortest? idk; or i'll just bug, and blame the government
                // because that's easier

                if (destIdx < 0) {
                    // (here, again, the old python server used to warn if we
                    // couldn't find the stop we're supposed to insert something
                    // relative to; but since our mot is run by toddlers who
                    // care more about highways to illegal settlements in
                    // occupied palestine than providing functioning public transit,
                    // this is such a common occurrence that even debug logging it
                    // feels utterly useless)
                    winston.debug(`tried adding stop relative to stop on on route; route_id=${route_id}, ${JSON.stringify(change)}, alert_id=${alertRaw.id}, trip_id=${representativeTripId}`);

                    continue;
                }

                if (!change.is_before) {
                    destIdx += 1;
                }

                updatedStopSeq.splice(destIdx, 0, [change.added_stop_id, true]);
                winston.debug(`added stop ${change.added_stop_id} to route ${route_id} at index ${destIdx}`);

                if (mutAllStopIdsSet) {
                    mutAllStopIdsSet.add(change.added_stop_id);
                }
            }
        }
    }

    if (mutAllStopIdsSet) {
        deletedStopIds.forEach(s => mutAllStopIdsSet.add(s));
    }

    return {
        representativeDate,
        representativeTripId,
        rawStopSeq,
        updatedStopSeq,
        deletedStopIds
    };
}

function removeStopFromUpdatedStopSeq(
    updatedStopSeq: [string, boolean][],
    stopId: string
) {
    let didRemove = false;
    let nextIndex;
    while ((nextIndex = updatedStopSeq.findIndex(([s, _]) => s === stopId)) >= 0) {
        updatedStopSeq.splice(nextIndex, 1);
        didRemove = true;
    }

    return didRemove;
}

async function getFallbackShape(rawStopSeq: string[], gtfsDbApi: GtfsDbApi) {
    const stopData = await gtfsDbApi.getStopsForMap(rawStopSeq);

    const result: [number, number][] = [];
    for (const stopId of rawStopSeq) {
        const stop = stopData[stopId];
        if (!stop) continue;

        result.push([stop.stop_lon, stop.stop_lat]);
    }
    return result;
}

const ROUTE_DESC_DIR_ALT_REGEX = /^[^-]+-([^-]+)-([^-]+)$/g;

function labelLineChangesHeadsignsForDirectionAndAlternative(
    mut_lineChanges: RouteChangeForApi[]
) {
    const headsignToDirAltPairs: Record<string, [string, string][]> = {};

    const headsignDirAlt_perChange: [string, [string, string]][] = [];

    for (const chg of mut_lineChanges) {
        const dirAltPairs = headsignToDirAltPairs[chg.to_text]
            ?? (headsignToDirAltPairs[chg.to_text] = []);
        
        // using matchAll here because i want to avoid Weird JS Regex Behavior(tm)
        const regexMatch = [...chg.route_desc.matchAll(ROUTE_DESC_DIR_ALT_REGEX)][0];
        const pair: [string, string] = [
            regexMatch?.[1] ?? "",
            regexMatch?.[2] ?? ""
        ];

        dirAltPairs.push(pair);
        headsignDirAlt_perChange.push([chg.to_text, pair]);
    }

    const dirAltNames_perChange = labelHeadsignsForDirectionAndAlternative(
        headsignToDirAltPairs,
        headsignToDirAltPairs,
        headsignDirAlt_perChange
    );

    for (const [chg, [dirName, altName]] of zip(mut_lineChanges, dirAltNames_perChange)) {
        if (dirName) chg.dir_name = dirName;
        if (altName) chg.alt_name = altName;
    }
}

function *labelHeadsignsForDirectionAndAlternative(
    byAlt_headsignToDirAltPairs: Record<string, [string, string][]>,
    byDir_headsignToDirAltPairs: Record<string, [string, string][]>,
    headsignDirAlt_list: [string, [string, string]][],
    labelDirsPerAlt = false
): IterableIterator<[string|null, string|null]> {
    for (const [headsign, dirAltPair] of headsignDirAlt_list) {
        const dupsByAlt = byAlt_headsignToDirAltPairs[headsign] ?? [];
        const dupsByDir = byDir_headsignToDirAltPairs[headsign] ?? [];

        let dirName: string|null = null;
        let altName: string|null = null;

        if (dupsByAlt.length <= 1 && dupsByDir.length <= 1) {
            // no duplicates for this headsign! yay upside down smiley
            yield [dirName, altName];
            continue;
        }

        const [dirId, altId] = dirAltPair;

        if (dupsByDir.some(([dir, alt]) => dir !== dirId && (!labelDirsPerAlt || alt === altId))) {
            // if there's any dups with a different direction id

            // in some distant nebulous future, i could try giving actual names
            // to the directions and alternatives; but not today, not quite yet
            // bukra fil mishmish
            dirName = "" + getNumberForDirection(labelDirsPerAlt, dirId, altId, dupsByDir);
        }

        if (altId !== "#" && altId !== "0" && dupsByAlt.some(([_, alt]) => alt !== altId)) {
            // if there's any dups with a different alternative id
            // (and also this isn't the main alternative)

            // remember what i said about actual names for directions?
            // well, same for the alternatives
            // i mean, like, how the heck does one even approach this problem???
            // we're given basically zero computer readable information that
            // summarizes the differences between directions/alternatives!
            // i guess if someone was eager enough, they COULD go through the
            // stop_seq of a representative trip, but then, how do you turn that
            // into not only user-readable info, but user-useful info, that isn't
            // too long to fit in the ui?????? can't just dump stop names/city names,
            // and detecting street names would be absolutely hellish

            // so uhm,, yeah
            // numbers it is for now

            // again possibly the slowest most inefficient blah blah blah
            // note: a !== "#" && a !== "0"] cause we don't care about the main
            // alternative here

            // i want to display: Towards A, Towards A (Alt 1), Towards A (Alt 2)
            // and not quite:     Towards A, Towards A (Alt 2), Towards A (Alt 3)

            const alternatives = dupsByAlt
                .filter(([_, a]) => a !== "#" && a !== "0")
                .map(([_, a]) => a);
            inPlaceSortAndUnique(alternatives);

            if (alternatives.length === 1) {
                // but also we want to display: Towards A, Towards A (Alt)
                // and not quite:               Towards A, Towards A (Alt 1)
                // because "1" doesn't make sense when there's just the one
                altName = "#"
            } else {
                altName = "" + (alternatives.indexOf(altId) + 1);
            }
        }

        yield [dirName, altName];
    }
}

function getNumberForDirection(
    labelDirsPerAlt: boolean,
    dirId: string,
    altId: string,
    dupsByDir: [string, string][]
) {
    // possible the slowest most inefficient way to do this but as
    // stated earlier, yours truly is truly bad at algo

    const filteredPairs =
        labelDirsPerAlt
            ? dupsByDir.filter(([_, a]) => a === altId)
            : dupsByDir;
    
    const justDirections = filteredPairs.map(([d, _]) => d);
    inPlaceSortAndUnique(justDirections);

    return justDirections.indexOf(dirId) + 1;
}

function boundingBoxForStops(
    stopIds: Iterable<string>,
    stops_for_map: Record<string, StopForMap>
): MapBoundingBox {
    /**
     * get bounding box of affected stops, for setting the maps' bounding box
     */

    let min_lon = Infinity;
    let min_lat = Infinity;
    let max_lon = -Infinity;
    let max_lat = -Infinity;

    for (const stopId of stopIds) {
        const stop = stops_for_map[stopId];
        if (!stop) continue;

        if (min_lon > stop.stop_lon) {
            min_lon = stop.stop_lon;
        }
        if (min_lat > stop.stop_lat) {
            min_lat = stop.stop_lat;
        }
        if (max_lon < stop.stop_lon) {
            max_lon = stop.stop_lon;
        }
        if (max_lat < stop.stop_lat) {
            max_lat = stop.stop_lat;
        }
    }

    return {
        min_lon,
        min_lat,
        max_lon,
        max_lat
    };
}
